### PR TITLE
Unify TwoIndex

### PR DESCRIPTION
**Pull request by chillenzer**
Hi,

as discussed in the meeting. Needs an update on `testGenerators`.

Best,
Julian

### CODE REVIEW

**Comment by LupoA** [comment]
Test_hmc_Sp_Wilson2ASFermionGauge is not compiling for me!

File: Grid/qcd/utils/GaugeGroupTwoIndex.h [hunk 1]
```diff
@@ -25,90 +25,139 @@
...
+class GaugeGroupTwoIndex : public GaugeGroup<ncolour, group_name> {
+ public:
+  // The chosen convention is that we are taking ncolour to be N in SU<N> but 2N
+  // in Sp(2N). ngroup is equal to N for SU but 2N/2 = N for Sp(2N).
+  static_assert(std::is_same<group_name, GroupName::SU>::value or
+                    std::is_same<group_name, GroupName::Sp>::value,
+                "ngroup is only implemented for SU and Sp currently.");
+  static const int ngroup =
+      std::is_same<group_name, GroupName::SU>::value ? ncolour : ncolour / 2;
+  static const int Dimension = ngroup * (ncolour + S) / 2;
```
Comment (LupoA) [inline Grid/qcd/utils/GaugeGroupTwoIndex.h]: I think here is the problem! The dimension of the representation, for Sp(ncolour), should be 
`ncolour (ncolour + S)/2  + S`
while now it reads
`ngroup (ncolour + S) / 2 = ncolour (ncolour + S) / 4`

**Review by LupoA — DISMISSED** [review]

File: Grid/qcd/utils/GaugeGroupTwoIndex.h (hunk 1 above)
Comment (LupoA) [reply Grid/qcd/utils/GaugeGroupTwoIndex.h]: even by adjusting this Test_hmc_Sp_Wilson2ASFermionGauge is not compiling

**Comment by LupoA** [comment]
@chillenzer I've committed changes since they were small, however my definition of the dimension only accounts for Sp and SU... could you make it a bit more pretty? 😇 thanks

Currently it is
```
static const int Dimension =
      std::is_same<group_name, GroupName::SU>::value ? ncolour * (ncolour + S) / 2 : ngroup * (ncolour + S) + S;
```

The real problem is that the HMC gives now bad physics. I suspect the generators are no longer correct!

**Comment by LupoA** [comment]
alright the problem was turning SUnTwoIndex into GaugeGroup instead of Sp2nTwoIndex since the former had some hardcoded crap... should be a quick fix

**Comment by chillenzer** [comment]
Great! Thanks so much!
Could you include the test that verifies the behaviour as an automated test? (If you run into any trouble, we can do that together and/or you can upload enough for me to figure out what you did precisely and try myself to automate this.)

**Comment by chillenzer** [comment]
> @chillenzer I've committed changes since they were small, however my definition of the dimension only accounts for Sp and SU... could you make it a bit more pretty? innocent thanks
> 
> Currently it is
> 
> ```
> static const int Dimension =
>       std::is_same<group_name, GroupName::SU>::value ? ncolour * (ncolour + S) / 2 : ngroup * (ncolour + S) + S;
> ```
> 
> The real problem is that the HMC gives now bad physics. I suspect the generators are no longer correct!

Sorry, I'm not sure what to do about this. We could go for
```
static const int Dimension = ncolour * (ncolour + S) / 2 +  (std::is_same<group_name, GroupName::Sp>::value ? S :  0);
```
or we could define a variable to shorten the `std::is_same...` stuff but other than that I don't see anything. I've ran into the same problem about using a simple ternary operator and thereby neglecting other possibilities but the `static_assert` in line 39 should warn anyone who attempts to do something else.

**Comment by LupoA** [comment]
> Great! Thanks so much!
> Could you include the test that verifies the behaviour as an automated test? (If you run into any trouble, we can do that together and/or you can upload enough for me to figure out what you did precisely and try myself to automate this.)

I just run 

`./Test_hmc_Sp_Wilson2ASFermionGauge --grid 4.4.4.4 --Trajectories 1 --Thermalizations 0 --StartingType ColdStart`

before and after changes and check dH and Plaquette at the end of the trajectory... maybe a bash/python script is enough for that?

**Comment by chillenzer** [comment]
Okay. I'll cook something up.

**Comment by LupoA** [comment]
Considering that we are almost done changing files I am also happy to check by eye in the next PR... Niccolò will do this extensively once he runs again the code to reproduce his plots, but if it's quick feel free to write a script

**Comment by LupoA** [comment]
I see... my only (mild) concern is for when SO(N) is implemented but we can definitely not worry about it. Could we have a dimension_setter struct or class that we specialise which each gauge group? I don't know if it even makes sense but something that would be used like:
`Dimension = dimension_setter<group_name>::this_dimension `
`GaugeGroupTwoIndex<ncolour, S, SU>::dimension_setter { this_dimension = ... }`
`GaugeGroupTwoIndex<ncolour, S, Sp>::dimension_setter { this_dimension = ... }`

**Comment by chillenzer** [comment]
That is the long-term perspective. But for us it's totally fine to have it that way and once someone starts implementing something else, they are clearly warned by the `static_assert`.

**Comment by chillenzer** [comment]
Two questions:
* When I run that exact command, I get a lot of `Force average: -nan` as well as `H = -nan` and `Total H after trajectory  = -nan  dH = -nan` although everything else seems to have values. That doesn't sound right. But it's true before and after my change.
* Do you have an idea of how reproducible the numbers are?

**Comment by LupoA** [comment]
ah shit... something might have been broken in the previous PR. If I remember you compile with Nc=2 while I use Nc=4 and since I lack the patience to check the Nc dependence I might have unwillingly done something that doesn't work for other Nc's.

I'll look into that... to be sure, what is your config line?

the numbers should be reproducible exactly since the random seeds are fixed in the test

**Comment by chillenzer** [comment]
```
../configure --enable-simd=AVX --enable-comms=mpi-auto --enable-Nc=2 CXX=... CC=... CXXFLAGS=-Wno-macro-redefined --no-create --no-recursion
```

**Comment by chillenzer** [comment]
Okay. Added some very simple automation for the test. I assumed that the current Nc=4 version is correct and used the value that it yields. (That's not ideal but I didn't have an independent source at hand.) It only checks the pseudofermion action because that was already lying around there but it feels highly unlikely that you can reproduce that to 8 significant digits and still do something wrong that could have been detected with this test.

We should do some things to make this test more robust though:
* Add correct values for other Nc once the other Ncs are running correctly.
* Find a nicer way than just crashing if an untested Nc is passed. You could probably do some automake/autoconfigure magic to only run this test if a pertinent Nc is used, or you could run this test but not do the assertion if we don't know the correct value for Nc. Whatever we choose, we should polish this a bit.

(Also: We are, of course, testing something pseudorandom here that would change whenever something of the underlying implementation details changes. That's also not ideal.)

**Comment by LupoA** [comment]
sorry nothing broke: the 2AS representation for SU(2) is just the singlet so nothing wrong with you getting nannies for Nc=2... actually a healthy sign 😅

**Comment by LupoA** [comment]
I am not sure how I feel about the automated tests for the HMC... tests so far are checking for specific functions. We then make sure the correct physics is produced by running a full simulation. With this idea, an HMC test like the proposed one should be redundant since more specific tests, targeting single functions, would capture any failure (ideally). In fact, the trouble with this PR was not with the HMC itself but the lack of a test on the generators. This will be implemented soon and a test will scan several values of Nc such as Test_sp2n_lie_gen does for the fundamental.

In short: I'd discard the automated HMC test and merge this PR as testGenerators() is ready to go

**Comment by LupoA** [comment]
I've adjusted the test for the generators in the 2as. They fail for Sp(Nc) with Nc>4 but this seems to be a pre-refactoring issue... no problem with SU(N)

**Comment by chillenzer** [comment]
Okay. I discussed with Ed yesterday and my quick test is indeed very brittle. I reverted the commit for now. We still agreed that it would be a good idea to have such a test but the details need more time (which we might or might not be willing to invest in the future).

**Comment by LupoA** [comment]
It seems there is a problem with the mapping of the two indices into a single one

**Comment by LupoA** [comment]
The function base in GaugeGroupTwoIndex only works for Sp4

**Comment by LupoA** [comment]
now it should work for every Nc 🤤you'll see I've used an if statement in the base function to discriminate between sp and SU... if you spot anything more elegant please change!

the 2index-symmetric representation is still broken for Sp but that is pre-refactoring issue. I'd say let's give myself until tomorrow to fix it and if I can't we move on for the time being.
